common/persistence/sql/sqlplugin/mysql/visibility.go

```diff
@@ -80,7 +80,9 @@ var errCloseParams = errors.New("missing one of {closeStatus, closeTime, history
 // its left as such and no update will be made
 func (mdb *db) InsertIntoVisibility(ctx context.Context, row *sqlplugin.VisibilityRow) (sql.Result, error) {
 	row.StartTime = mdb.converter.ToMySQLDateTime(row.StartTime)
-	return mdb.conn.ExecContext(ctx,
+	dbShardID := sqlplugin.GetDBShardIDFromDomainID(row.DomainID, mdb.GetTotalNumDBShards())
```

Comment:
Shard by domain ID is not a good implementation, because it will cause large partition issue

Reply:
I don't think we have better option based on the schema and our query pattern: https://github.com/uber/cadence/blob/master/schema/mysql/v57/visibility/schema.sql

It's okay for here. A Cadence cluster with large scale should use advanced visibility anyway. We just need to do our best here based on the current architecture

Reply:
what about `run_id` or `domain_id, run_id`?

Reply:
It will not work. Because visibilityStore needs to query by workflowID or closeStatus etc. 
The indexes are also built with that pattern: https://github.com/uber/cadence/blob/master/schema/mysql/v57/visibility/schema.sql#L19

If the partition is based on runID or `domainID+runID`(no difference because runID is already random), a single query/index will need to be sent to all shards and then merge back.

Also, Cassandra uses domainID as partitionKey as well. (domainPatition is calculated by domainID). https://github.com/uber/cadence/blob/master/schema/cassandra/visibility/schema.cql